salt/state.py

```diff
@@ -1261,6 +1294,15 @@ def compile_high_data(self, high):
                                   not isinstance(val, six.string_types)):
                                 # Invalid name, fall back to ID
                                 chunk[key] = name
+                            elif ('__sls__' in body and
+                                  '__env__' in body and
+                                  key in STATE_REQUISITE_KEYWORDS):
+                                _val = type(val)()
```

Comment:
while this is a very cool way to do it, I would just make `_val` a dict, this is because you are explicitly reconstructing _val as a dict, so using this could give you a value which could cause the reconstruction to stack trace.